ipn/ipnlocal/profiles.go

```diff
@@ -0,0 +1,467 @@
+// Copyright (c) 2022 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package ipnlocal
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/rand"
+	"runtime"
+	"sync"
+	"time"
+
+	"golang.org/x/exp/slices"
+	"tailscale.com/ipn"
+	"tailscale.com/types/logger"
+	"tailscale.com/util/strs"
+)
+
+// ProfileManager is a wrapper around a StateStore that manages
+// multiple profiles and the current profile.
+type ProfileManager struct {
+	store ipn.StateStore
+	logf  logger.Logf
+
+	// Lock order: LocalBackend.mu, then pm.mu.
+	mu             sync.Mutex // guards following
+	currentUserID  string     // only used on Windows
+	knownProfiles  map[ipn.ProfileID]*ipn.LoginProfile
+	currentProfile *ipn.LoginProfile
+	prefs          ipn.PrefsView
+
+	// isNewProfile is a sentinel value that indicates that the
+	// current profile is new and has not been saved to disk yet.
+	// It is reset to false after a call to SetPrefs with a filled
+	// in LoginName.
+	isNewProfile bool
+}
+
+// CurrentUser returns the current user ID. It is only non-empty on
+// Windows where we have a multi-user system.
```

Comment:
Are we going to want something like this for the macsys build too, which has a shared storage for all users? CC @nickoneill

Reply:
Seems reasonable. I haven't experimented with multi-user macOS installs in a while and maybe never with macsys builds, but prefixing the state files on macsys with a userid would be a nice improvement for anyone who uses more than one login.

Reply:
that is going to need to happen at a different layer than the ProfileManager as we need to do UserID detection and have that plumbed in. It's not the same as using `os.GetUid()` 

See https://github.com/tailscale/tailscale/blob/446057d6134055ab60d96f4ceee67c9d89e812f3/ipn/ipnserver/server.go#L138-L185